File: ventures/0016-hedgefonds-simulation-echte-weltwirtschaft/befunde/messung-0105/baum_tot/ventures/0016-hedgefonds-simulation-echte-weltwirtschaft/messung-stack/java/Schritt.java
import java.math.BigInteger;

/**
 * Deterministischer Festkomma-Weltschritt, Messfassung fuer den Stackvergleich.
 *
 * Ganzzahlarithmetik durchgehend, kein Gleitkommatyp -- auch nicht in der
 * Zeitmessung. Java rundet die Ganzzahldivision und den Restoperator gegen null
 * ab; das ist genau die von der Aufgabe verlangte Abschneiderichtung.
 */
public final class Schritt {

    private static final int N = 64;
    private static final int SCHRITTE = 1_000_000;
    private static final long UNTERE_GRENZE = -1_000_000_000_000L;
    private static final long OBERE_GRENZE = 1_000_000_000_000L;
    private static final long NENNER = 10_000L;

    private Schritt() {
    }

    public static void main(String[] args) {
        long[] z = new long[N];
        for (int i = 0; i < N; i++) {
            z[i] = 1_000_000L + (long) i * 37L;
        }

        long begonnen = System.nanoTime();
        for (int s = 0; s < SCHRITTE; s++) {
            weltschritt(z);
        }
        long gedauert = System.nanoTime() - begonnen;

        System.out.println("pruefsumme=" + pruefsumme(z));
        System.out.println("nanosekunden_je_schritt=" + gedauert / SCHRITTE);
        System.out.println("zustand0=" + z[0]);
    }

    /** Ein Weltschritt, i aufsteigend, an Ort und Stelle. */
    private static void weltschritt(long[] z) {
        for (int i = 0; i < N; i++) {
            long nachbar = z[(i + 17) & (N - 1)];
            long alt = z[i];
            long roh = malGeteilt(alt, 9_512L + nachbar % 977L, NENNER);
            z[i] = klemme(roh + nachbar / 1024L - alt / 4096L, UNTERE_GRENZE, OBERE_GRENZE);
        }
    }

    /**
     * a * b / c ohne Ueberlauf im Zwischenergebnis, gerundet auf halbe Betraege
     * von null weg. c ist positiv.
     *
     * Passt das Produkt in 64 Bit -- was bei geklemmtem Zustand immer der Fall
     * ist --, laeuft der schnelle Weg; sonst rechnet BigInteger die vollen
     * 128 Bit. Die Erkennung vergleicht das hohe Wort mit der Vorzeichenmaske
     * des niedrigen.
     */
    private static long malGeteilt(long a, long b, long c) {
        long tief = a * b;
        long hoch = Math.multiplyHigh(a, b);
        if (hoch == (tief >> 63)) {
            return geteiltGerundet(tief, c);
        }
        BigInteger produkt = BigInteger.valueOf(a).multiply(BigInteger.valueOf(b));
        BigInteger nenner = BigInteger.valueOf(c);
        BigInteger betrag = produkt.abs();
        BigInteger ganz = betrag.divide(nenner);
        BigInteger rest = betrag.subtract(ganz.multiply(nenner));
        if (rest.shiftLeft(1).compareTo(nenner) >= 0) {
            ganz = ganz.add(BigInteger.ONE);
        }
        if (produkt.signum() < 0) {
            ganz = ganz.negate();
        }
        return ganz.longValueExact();
    }

    /** n / c auf halbe Betraege von null weg gerundet, c positiv. */
    private static long geteiltGerundet(long n, long c) {
        long betrag = n < 0 ? -n : n;
        long ganz = betrag / c;
        long rest = betrag - ganz * c;
        if (rest >= c - rest) {
            ganz++;
        }
        return n < 0 ? -ganz : ganz;
    }

    private static long klemme(long x, long unten, long oben) {
        if (x < unten) {
            return unten;
        }
        if (x > oben) {
            return oben;
        }
        return x;
    }

    /** summe(z[i] * (i + 1)) mit Ueberlaufumbruch, danach mod 2^63 - 1. */
    private static long pruefsumme(long[] z) {
        long summe = 0L;
        for (int i = 0; i < N; i++) {
            summe += z[i] * (long) (i + 1);
        }
        return summe % Long.MAX_VALUE;
    }
}

File: ventures/0016-hedgefonds-simulation-echte-weltwirtschaft/befunde/messung-0105/baum_tot/ventures/0016-hedgefonds-simulation-echte-weltwirtschaft/messung-stack/rust/schritt.rs
// Deterministischer Festkomma-Weltschritt fuer die Stackmessung 0016.
// Freistehende Datei, keine Fremdbibliothek: rustc -O schritt.rs
//
// Ganzzahldivision und Rest schneiden in Rust gegen null ab -- genau die
// Vorgabe aus AUFGABE.md. Deshalb wird `/` und `%` unveraendert benutzt.

use std::time::Instant;

const N: usize = 64;
const SCHRITTE: u64 = 1_000_000;
const UNTERGRENZE: i64 = -1_000_000_000_000;
const OBERGRENZE: i64 = 1_000_000_000_000;

/// `a * b / c` ohne Ueberlauf im Zwischenergebnis (128 Bit),
/// gerundet auf halbe Betraege von null weg.
fn mal_geteilt(a: i64, b: i64, c: i64) -> i64 {
    let zaehler = (a as i128) * (b as i128);
    let nenner = c as i128;
    let negativ = (zaehler < 0) != (nenner < 0);
    let betrag_zaehler = if zaehler < 0 { -zaehler } else { zaehler };
    let betrag_nenner = if nenner < 0 { -nenner } else { nenner };
    // floor((|z| + |n|/2) / |n|), exakt auch bei ungeradem Nenner:
    let betrag = (2 * betrag_zaehler + betrag_nenner) / (2 * betrag_nenner);
    (if negativ { -betrag } else { betrag }) as i64
}

/// Begrenzt `x` auf `[unten, oben]`.
fn klemme(x: i64, unten: i64, oben: i64) -> i64 {
    if x < unten {
        unten
    } else if x > oben {
        oben
    } else {
        x
    }
}

/// Ein Weltschritt: `i` aufsteigend, an Ort und Stelle.
fn weltschritt(z: &mut [i64; N]) {
    for i in 0..N {
        let nachbar = z[(i + 17) % N];
        let roh = mal_geteilt(z[i], 9_512 + (nachbar % 977), 10_000);
        z[i] = klemme(
            roh + (nachbar / 1024) - (z[i] / 4096),
            UNTERGRENZE,
            OBERGRENZE,
        );
    }
}

/// `summe(z[i] * (i + 1))` mit Ueberlaufumbruch in 64 Bit.
fn pruefsumme(z: &[i64; N]) -> i64 {
    let mut summe: i64 = 0;
    for i in 0..N {
        summe = summe.wrapping_add(z[i].wrapping_mul(i as i64 + 1));
    }
    summe
}

fn main() {
    let mut z = [0i64; N];
    for i in 0..N {
        z[i] = 1_000_000 + (i as i64) * 37;
    }

    let start = Instant::now();
    for _ in 0..SCHRITTE {
        weltschritt(&mut z);
    }
    let nanosekunden = start.elapsed().as_nanos();

    println!("pruefsumme={}", pruefsumme(&z));
    println!(
        "nanosekunden_je_schritt={}",
        nanosekunden / (SCHRITTE as u128)
    );
    println!("zustand0={}", z[0]);
}
